Fill transparent areas of image before hashing

smason opened the pull request:
PIL and Wand code fixing issue #3 and the start of a test framework

To run test code, execute:

```python -m unittest```

The code was getting a bit verbose so I restructured things a bit, hope it's OK.

smason committed:
add a (failing) testcase for difference in transparent part of image

smason committed:
add support for filling transparent areas

and update tests for compatibility with Wand as it uses the same Image
class ane

smason committed:
add code to support transparency filling to Wand image objects

restructure code as it was getting a bit fiddly

smason committed:
get grayscale blob directly from Wand

asking for RGB then slicing off everything other than blue seems
unnecessary

smason committed:
remove fill_transparent function

oops, forgot this earlier!

benhoyt commented on dhash.py:
```diff
@@ -29,7 +30,36 @@
+
+def get_grays_wand(image, width, height, fill_color='white'):
+    # we don't want to mutate the caller's image
+    image = image.clone()
```
Should this be using `with image.clone() as image:` to enable proper wand/ImageMagick resource cleanup?

benhoyt commented on dhash.py:
```diff
@@ -29,7 +30,36 @@
+
+    image.resize(width, height)
+
+    blob = image.make_blob(format='GRAY')
```
Is doing this the same as what I had with `format='RGB'` and then `blob[::3]`? Seems to be in my small doctests against `dhash-test.jpg`, but not 100% sure.

smason replied:
can't remember now, sorry!

at a quick guess, `blob[::3]` is just pulling out the red channel as it's conventionally the first

benhoyt commented on test.py:
```diff
@@ -0,0 +1,35 @@
+from unittest import TestCase
```
Good idea to add this rather than trying to stuff more things in doctests. Can you also please convert the two dhash-test.jpg doctests (in get_grays and dhash_row_col) to unit tests in this file? I think the others (that don't use wand/PIL) can stay as doctests.

Then it'd be great if just running this script would run the unittests. I think this is as simple as adding this to the bottom of test.py.

```python
if __name__ == '__main__':
    unittest.main()
```

smason replied:
I always seem to use `pytest` as a runner, but have made this change anyway

benhoyt reviewed (commented):
Thanks for this! Sorry for taking so long to get to reviewing. I've added a couple of questions/comments.

For bonus points (though I can do this too), it'd be great to:

* add `tests_require=['Pillow', 'wand']` to setup.py
* change to using `from setuptools import setup` in setup.py to enable "python setup.py test"
* add `.travis.yml` and `appveyor.yml` so GitHub can automatically run tests on commit

benhoyt commented:
Hi @smason, any interest in addressing the above and getting this over the goal line? I know it's been a few years!

smason committed:
Merge remote-tracking branch 'upstream/master' into fill-transparency

smason committed:
switch to using setuptools and add tests_require

smason committed:
use context manager around clone to allow safe cleanup

smason committed:
allow tests to be run via: python test.py

smason committed:
get simplified doctest working

mostly failed because of missing import, but os stuff feels a bit over
the top for a doctest

smason committed:
move doctests to test.py

smason commented:
Hi, not sure what happened here!  I've made most of the changes you commented on.

You made a comment about using Travis, would GitHub's actions be the better option now?

benhoyt commented on dhash.py:
```diff
@@ -33,26 +34,55 @@
 
 
-def get_grays(image, width, height):
+def get_grays_pil(image, width, height, fill_color='white'):
```
Can we please rename these to `_get_grays_pil` and `_get_grays_wand` as I'd prefer they weren't part of the public API?

benhoyt commented on dhash.py:
```diff
@@ -80,12 +110,6 @@ def dhash_row_col(image, size=8):
-    >>> format(row, '016b')
-    '0100101111010001'
-    >>> format(col, '016b')
-    '0101001111111001'
```
I think we can leave this doctest in as it doesn't depend on external image data, and it's a nice little example of how to use it with a list input.

benhoyt commented on test.py:
```diff
@@ -0,0 +1,64 @@
+IMGDIR = path.dirname(__file__)
+
+
+def pilToWand(image, format='png'):
```
Naming nit: let's pleas use `pil_to_wand` naming for this for consistency.

smason replied:
yup, good to stick to this

benhoyt commented on test.py:
```diff
@@ -0,0 +1,64 @@
+        expected = [
+            93, 158, 210, 122, 93, 77, 74, 74, 77,
+            95, 117, 122, 111, 92, 74, 81, 80, 77,
+        ]  # fmt: skip
```
We may have to make this allow variation by +/-1, see [here](https://github.com/benhoyt/dhash/commit/730ce501ab1ca68c49399e407ead8844dd932a57#diff-99b846d6215ee4329114b53913fed329569ac1f94e31f5dc5330a216d383195eR48).

smason replied:
oops, misread the intent of the code.  switched to using `delta=1` which should do this

benhoyt reviewed (commented):
Thanks! Just a couple of small comments. Yeah, you're right, we're using GitHub Actions now, so please add running the new `test.py` tests to `.github/workflows/ci.yml` as well.

smason committed:
run github actions via pytest, add coverage

pytest has nice coverage metrics, could turn this into a nice badge
if wanted later

smason committed:
also test doctests

smason committed:
fix naming convention and tests

smason committed:
move test back to doctest

smason commented:
think I've done those now

added pytest to the runner as it gives nice coverage output, also added doctests to the test runner.

hope that's all okay!

smason committed:
fuzzyness needed here as well

only seems to be needed for Python 2.7

benhoyt committed:
Pull in fix for #9 (deprecated PIL.Image.ANTIALIAS)

This was missed during the merge of the work from PR #11.

Also tweak some formatting

benhoyt committed:
Remove duplicated code; revert some formatting changes

benhoyt committed:
Fix tests on Python 2.7; add wand test case

benhoyt commented:
Thanks for this! I added a couple more changes: fixing the tests on Python 2.7, and pulling in something forgotten from the merge (the deprecated `PIL.Image.ANTIALIAS` constant).

smason commented:
Looks good, thanks!

I haven't used Python 2.7 for a long time now, any idea how much use it's actually getting?

benhoyt commented:
@smason Neither have I, but https://pypistats.org/packages/dhash shows that dhash is still getting ~20 downloads per day on Python 2.7 (compared to ~400 per day on Python 3.x).
